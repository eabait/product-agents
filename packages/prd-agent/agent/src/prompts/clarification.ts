/**
 * Clarification Prompt
 * 
 * Analyzes user input to determine if sufficient context exists for PRD generation
 * and generates targeted questions to gather missing information.
 */

export function createClarificationPrompt(userMessage: string): string {
  const trimmedMessage = userMessage.trim()

  return [
    'You evaluate whether the following product request has enough detail to generate a meaningful PRD.',
    '',
    `Request: "${trimmedMessage.length > 0 ? trimmedMessage : 'No specific user request provided.'}"`,
    '',
    'Determine if clarification is required:',
    '- Minimum viable information covers the product concept, primary user type, and essential functionality.',
    '- Prefer making reasonable industry assumptions over asking obvious follow-up questions (e.g., payments for e-commerce, task tracking for project tools).',
    '',
    'Questioning strategy:',
    '- Ask only critical questions when a core concept, users, or primary functionality is missing.',
    '- Optional clarifications should be limited to high-impact business context, constraints, or success criteria.',
    '',
    'Confidence guidelines (0-100):',
    '- 80-100 → proceed without clarification.',
    '- 60-79 → proceed but include up to two critical questions.',
    '- 40-59 → ask targeted questions that resolve fundamental gaps.',
    '- 0-39 → request clarification before drafting the PRD.',
    '',
    'Return JSON with keys:',
    '- needsClarification (boolean)',
    '- confidence (number 0-100)',
    '- missingCritical (array of strings describing missing essentials)',
    '- questions (array of critical questions only)',
    '',
    'Output rules:',
    '- Return strict JSON only. No prose, explanations, or code fences.'
  ].join('\n')
}
