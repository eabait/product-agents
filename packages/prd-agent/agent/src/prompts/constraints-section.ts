import { SectionWriterInput } from '../section-writers/base-section-writer'
import {
  buildAnalysisSummaryBlock,
  buildExistingSectionBlock,
  buildUserContextBlock,
  formatExistingItemsList,
  formatReturnJsonOnly,
  formatStructuredOutputRequirement
} from './prompt-helpers'

export function createConstraintsSectionPrompt(
  input: SectionWriterInput,
  contextAnalysis: any,
  existingConstraints: string[],
  existingAssumptions: string[]
): string {
  const lines: string[] = [
    'You are a product manager refining the Constraints section of a PRD. Apply minimal edits while preserving strong existing content.',
    '',
    buildUserContextBlock(input.message),
    buildAnalysisSummaryBlock(contextAnalysis, { includeConstraints: true }),
    ''
  ]

  const existingSection = buildExistingSectionBlock(input, 'constraints', { assumptions: true })
  if (existingSection) {
    lines.push(existingSection, '')
  }

  const currentConstraints = formatExistingItemsList('## Current Constraints', existingConstraints)
  if (currentConstraints) {
    lines.push(currentConstraints, '')
  }

  const currentAssumptions = formatExistingItemsList('## Current Assumptions', existingAssumptions)
  if (currentAssumptions) {
    lines.push(currentAssumptions, '')
  }

  lines.push(
    '## Instructions',
    '- Produce 2-6 constraints (non-negotiable limits) and 2-5 assumptions (items to validate).',
    '- Keep statements specific, actionable, and â‰¤ 200 characters.',
    '- When modifying or removing existing items, reference them directly.',
    '- Incorporate insights from the analysis summary and provided context.',
    '',
    formatStructuredOutputRequirement([
      'mode: "smart_merge" | "append" | "replace"',
      'constraints.operations: array of { action, reference, value?, rationale? } for targeted edits',
      'constraints.proposed: array of fully written constraints to add when appending',
      'assumptions.operations: same structure as constraints.operations',
      'assumptions.proposed: array of fully written assumptions for append actions',
      'summary: optional string explaining notable decisions'
    ]),
    '',
    formatReturnJsonOnly()
  )

  return lines.join('\n')
}
