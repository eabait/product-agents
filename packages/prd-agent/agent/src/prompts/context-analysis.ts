/**
 * Context Analysis Worker Prompts
 * 
 * Prompts for analyzing product requests and extracting key themes,
 * requirements, and constraints.
 */

export function createContextAnalysisPrompt(message: string, contextPayload?: any): string {
  const trimmedMessage = message.trim()
  const lines: string[] = [
    'You analyze the product request and extract planning signals for a PRD.',
    '',
    '## User Input',
    trimmedMessage.length > 0 ? trimmedMessage : 'No specific request provided.',
    ''
  ]

  const supplemental = formatSupplementalContext(contextPayload)
  if (supplemental.length > 0) {
    lines.push('## Provided Context')
    lines.push(...supplemental, '')
  }

  lines.push(
    '## Tasks',
    '- Themes: identify the main product goals or focus areas (string array).',
    '- Requirements: capture functional, technical, and UX requirements as arrays of concise statements.',
    '- Epic stories: include 3-8 epic level user stories (title + description) when possible.',
    '- MVP features: list essential launch capabilities.',
    '- Constraints: highlight technical, business, or compliance limitations.',
    '',
    '## Output JSON Shape',
    '- themes: string[]',
    '- requirements: { functional: string[]; technical: string[]; user_experience: string[]; epics?: { title: string; description: string }[]; mvpFeatures?: string[] }',
    '- constraints: string[]',
    '',
    '## Output Rules',
    '- Use empty arrays when information is unavailable.',
    '- Keep entries concise (≤ 200 characters).',
    '- Return strict JSON only with the keys above. No additional commentary or code fences.'
  )

  return lines.join('\n')
}

function formatSupplementalContext(contextPayload?: any): string[] {
  if (!contextPayload?.categorizedContext || contextPayload.categorizedContext.length === 0) {
    return []
  }

  const priorityOrder: Record<string, number> = { high: 3, medium: 2, low: 1 }

  const activeItems = contextPayload.categorizedContext
    .filter((item: any) => item?.isActive)
    .sort((a: any, b: any) => (priorityOrder[b?.priority] || 1) - (priorityOrder[a?.priority] || 1))

  const constraintLines = activeItems
    .filter((item: any) => item.category === 'constraint')
    .slice(0, 5)
    .map((item: any) => `- Constraint: ${item.title ?? 'Untitled'} — ${truncate(item.content)}`)

  const requirementLines = activeItems
    .filter((item: any) => item.category === 'requirement')
    .slice(0, 5)
    .map((item: any) => `- Requirement: ${item.title ?? 'Untitled'} — ${truncate(item.content)}`)

  const summary: string[] = []
  if (constraintLines.length > 0) {
    summary.push(...constraintLines)
  }
  if (requirementLines.length > 0) {
    summary.push(...requirementLines)
  }

  return summary
}

function truncate(value: unknown, maxChars = 240): string {
  if (typeof value !== 'string') {
    return String(value ?? '').slice(0, maxChars)
  }
  return value.length > maxChars ? `${value.slice(0, maxChars - 3)}...` : value
}
