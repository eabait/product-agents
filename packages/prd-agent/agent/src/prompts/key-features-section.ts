import { SectionWriterInput } from '../section-writers/base-section-writer'
import {
  buildAnalysisSummaryBlock,
  buildExistingSectionBlock,
  buildUserContextBlock,
  formatExistingItemsList,
  formatReturnJsonOnly,
  formatStructuredOutputRequirement
} from './prompt-helpers'

export function createKeyFeaturesSectionPrompt(
  input: SectionWriterInput,
  contextAnalysis: any,
  existingFeatures: string[]
): string {
  const lines: string[] = [
    'You are a product manager updating the Key Features section of a PRD. Honor strong existing items and make the smallest necessary changes.',
    '',
    buildUserContextBlock(input.message),
    buildAnalysisSummaryBlock(contextAnalysis, {
      includeEpics: true,
      includeMvpFeatures: true,
      includeConstraints: true
    }),
    ''
  ]

  const existingSection = buildExistingSectionBlock(input, 'keyFeatures')
  if (existingSection) {
    lines.push(existingSection, '')
  }

  const currentFeatures = formatExistingItemsList('## Current Feature Set', existingFeatures)
  if (currentFeatures) {
    lines.push(currentFeatures, '')
  }

  lines.push(
    '## Instructions',
    '- Produce 3-7 user-facing features. Each feature should include a short title and a value-focused description.',
    '- Prioritize items that address the identified requirements, epics, and MVP scope.',
    '- Keep language stakeholder-friendly (â‰¤ 220 characters per feature).',
    '- Prefer targeted updates or additions over full rewrites.',
    '',
    formatStructuredOutputRequirement([
      'mode: "smart_merge" | "append" | "replace"',
      'operations: array of { action, referenceFeature?, feature?, rationale? } to describe updates/removals',
      'proposedFeatures: array of fully written features to append when adding new items',
      'summary: optional string explaining notable changes'
    ]),
    '',
    formatReturnJsonOnly()
  )

  return lines.join('\n')
}
