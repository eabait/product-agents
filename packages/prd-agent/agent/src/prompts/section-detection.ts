import { summarizeExistingPrd } from './prompt-helpers'

const SECTION_DESCRIPTORS: Record<string, string> = {
  targetUsers: 'Who the product is for',
  solution: 'What we are building and how',
  keyFeatures: 'Core functionality and differentiators',
  successMetrics: 'How success is measured',
  constraints: 'Technical, business, or regulatory limits'
}

export function createSectionDetectionPrompt(message: string, existingPRD?: any): string {
  const trimmedMessage = message.trim()

  const lines: string[] = [
    'You review an edit request and decide which PRD sections require updates.',
    '',
    '## Available Sections',
    JSON.stringify(SECTION_DESCRIPTORS, null, 2),
    '',
    '## User Request',
    trimmedMessage.length > 0 ? trimmedMessage : 'No instruction provided.',
    ''
  ]

  const existingSnapshot = summarizeExistingPrd(existingPRD)
  if (existingSnapshot.length > 0) {
    lines.push(...existingSnapshot, '')
  }

  lines.push(
    '## Decision Process',
    '- Select only the sections that must change to satisfy the request.',
    '- If the request does not materially impact a section, leave it out.',
    '- Typical mapping:',
    '  • New or modified features → keyFeatures',
    '  • Changes to audience/personas → targetUsers',
    '  • Core solution changes or delivery approach → solution',
    '  • Measurement or KPI updates → successMetrics',
    '  • New limitations, dependencies, or compliance needs → constraints',
    '',
    '## Output JSON Shape',
    '- affectedSections: string[] containing section keys from the table above',
    '- reasoning: object where each key is a section and the value explains why it changes',
    '- confidence: "high" | "medium" | "low"',
    '',
    '## Output Rules',
    '- Be conservative: exclude sections unless there is a clear change driver.',
    '- Return strict JSON only. No prose, explanations, or code fences.'
  )

  return lines.join('\n')
}
