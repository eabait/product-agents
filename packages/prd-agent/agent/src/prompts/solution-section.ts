import { SectionWriterInput } from '../section-writers/base-section-writer'
import {
  buildAnalysisSummaryBlock,
  buildExistingSectionBlock,
  buildUserContextBlock,
  formatReturnJsonOnly,
  formatStructuredOutputRequirement
} from './prompt-helpers'

export function createSolutionSectionPrompt(input: SectionWriterInput, contextAnalysis: any): string {
  const lines: string[] = [
    'You are a product manager drafting the Solution Overview section of a PRD. Provide clear direction while staying high level.',
    '',
    buildUserContextBlock(input.message),
    buildAnalysisSummaryBlock(contextAnalysis, { includeConstraints: true }),
    ''
  ]

  const existingSection = buildExistingSectionBlock(input, 'solution')
  if (existingSection) {
    lines.push(existingSection, '')
  }

  lines.push(
    '## Instructions',
    '- Solution overview: describe WHAT is being built in 2 concise paragraphs or fewer.',
    '- Approach: outline HOW the team will execute (architecture, phases, or methodology) in 1 paragraph.',
    '- Address the identified requirements and constraints while keeping stakeholder-friendly language.',
    '- Avoid low-level implementation details or open questions.',
    '',
    formatStructuredOutputRequirement([
      'solutionOverview: string (120-600 characters, stakeholder friendly)',
      'approach: string (90-400 characters describing execution strategy)'
    ]),
    '',
    formatReturnJsonOnly()
  )

  return lines.join('\n')
}
