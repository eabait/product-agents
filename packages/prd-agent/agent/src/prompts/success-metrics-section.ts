import { SectionWriterInput } from '../section-writers/base-section-writer'

export function createSuccessMetricsSectionPrompt(input: SectionWriterInput, contextAnalysis: any): string {
  let prompt = `You are a product manager creating a Success Metrics section for a Product Requirements Document (PRD).

## Input Context:
**User Message:** ${input.message}

## Analysis Results:
**Themes:** ${contextAnalysis.themes?.join(', ') || 'None identified'}
**Requirements:** ${JSON.stringify(contextAnalysis.requirements || {}, null, 2)}
**Constraints:** ${contextAnalysis.constraints?.join(', ') || 'None identified'}

## Instructions:
Generate 2-4 key success metrics that will measure whether this product or feature achieves its goals.

**Guidelines:**
- Focus on outcome metrics, not activity metrics (e.g., "user retention" not "features built")
- Each metric must have a specific, measurable target (numbers, percentages, or concrete thresholds)
- Include realistic timelines for achieving targets (e.g., "within 6 months", "by Q2 2024")
- Balance business metrics (revenue, cost savings) with user metrics (satisfaction, adoption)
- Ensure metrics can actually be measured with available data/tools

**Metric Structure:**
- **Metric**: Clear name of what we're measuring
- **Target**: Specific, measurable goal (with numbers)
- **Timeline**: When we expect to achieve this target

**Example Success Metrics:**
- **Metric**: "Monthly Active Users", **Target**: "5,000+ MAU within 6 months", **Timeline**: "6 months post-launch"
- **Metric**: "Customer Support Ticket Reduction", **Target**: "30% decrease in support volume", **Timeline**: "3 months post-implementation" 
- **Metric**: "User Task Completion Rate", **Target**: "85% of users complete primary workflow", **Timeline**: "First month of usage"

**Avoid These Common Mistakes:**
- Vague targets like "increase engagement" or "improve satisfaction"
- Metrics that can't be realistically measured
- Too many metrics (focus on 2-4 most important)
- Activity metrics instead of outcome metrics

Generate metrics that clearly indicate product success and can guide decision-making.`

  // Add existing PRD context if available
  if (input.context?.existingPRD) {
    prompt += `\n\n## Existing PRD Context:
${JSON.stringify(input.context.existingPRD, null, 2)}

Ensure success metrics align with the target users, solution, and key features defined in the existing PRD.`
  }

  return prompt
}