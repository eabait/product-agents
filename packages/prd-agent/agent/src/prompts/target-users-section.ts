import { SectionWriterInput } from '../section-writers/base-section-writer'
import {
  buildAnalysisSummaryBlock,
  buildExistingSectionBlock,
  buildUserContextBlock,
  formatExistingItemsList,
  formatReturnJsonOnly,
  formatStructuredOutputRequirement
} from './prompt-helpers'

export function createTargetUsersSectionPrompt(
  input: SectionWriterInput,
  contextAnalysis: any,
  existingUsers: string[]
): string {
  const lines: string[] = [
    'You are a product manager updating the Target Users section of a PRD. Preserve strong personas and apply only necessary edits.',
    '',
    buildUserContextBlock(input.message),
    buildAnalysisSummaryBlock(contextAnalysis, { includeConstraints: true }),
    ''
  ]

  const existingSection = buildExistingSectionBlock(input, 'targetUsers')
  if (existingSection) {
    lines.push(existingSection, '')
  }

  const currentUsers = formatExistingItemsList('## Current Personas to Respect', existingUsers)
  if (currentUsers) {
    lines.push(currentUsers, '')
  }

  lines.push(
    '## Instructions',
    '- Produce 2-4 personas. Each persona should be 1-2 sentences describing who they are and their primary need.',
    '- Focus on specific user segments that align with the themes and requirements.',
    '- Avoid generic labels (e.g., "end users"); highlight context such as role, company size, or pain point.',
    '- Reference existing personas when updating or removing them.',
    '',
    formatStructuredOutputRequirement([
      'mode: "smart_merge" | "append" | "replace"',
      'operations: array of { action, referenceUser?, user?, rationale? } describing targeted edits',
      'proposedUsers: array of fully written personas to append',
      'summary: optional string explaining key decisions'
    ]),
    '',
    formatReturnJsonOnly()
  )

  return lines.join('\n')
}
