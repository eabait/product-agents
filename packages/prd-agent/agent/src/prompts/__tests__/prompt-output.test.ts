import { createClarificationPrompt } from '../clarification'
import { createContextAnalysisPrompt } from '../context-analysis'
import { createSectionDetectionPrompt } from '../section-detection'
import { createSolutionSectionPrompt } from '../solution-section'
import { createKeyFeaturesSectionPrompt } from '../key-features-section'
import { createSuccessMetricsSectionPrompt } from '../success-metrics-section'
import { createConstraintsSectionPrompt } from '../constraints-section'
import { createTargetUsersSectionPrompt } from '../target-users-section'
import { SectionWriterInput } from '../../section-writers/base-section-writer'

const contextAnalysisSample = {
  themes: ['Analytics enablement', 'Enterprise readiness'],
  requirements: {
    functional: ['Stream KPIs in real time', 'Provide historical comparisons'],
    technical: ['Integrate with Snowflake warehouse'],
    user_experience: ['Accessible dashboards meeting WCAG 2.1 AA'],
    epics: [
      {
        title: 'Real-time dashboards',
        description: 'Operations leaders view live performance trends.'
      }
    ],
    mvpFeatures: ['Live KPI dashboard', 'Alert subscription management']
  },
  constraints: ['Must comply with SOC 2 Type II']
}

const existingPrdSample = {
  sections: {
    targetUsers: {
      targetUsers: [
        'Operations managers at mid-market retail companies needing live visibility'
      ]
    },
    solution: {
      solutionOverview:
        'We currently provide scheduled analytics reports summarizing daily performance.',
      approach: 'Deliver enhancements in three phases aligned to quarterly releases.'
    },
    keyFeatures: {
      keyFeatures: [
        'Automated scheduled reports delivered via email',
        'Role-based dashboards for executives and managers'
      ]
    },
    successMetrics: {
      successMetrics: [
        {
          metric: 'Weekly Active Analysts',
          target: '60% of analysts log in weekly',
          timeline: 'Within two quarters of launch'
        }
      ]
    },
    constraints: {
      constraints: ['Dependent on data warehouse refresh time under five minutes'],
      assumptions: ['Data engineering team maintains ingestion pipeline health']
    }
  }
}

const baseInput: SectionWriterInput = {
  message: 'Add real-time analytics for enterprise customers and update the launch strategy.',
  context: {
    existingPRD: existingPrdSample,
    contextPayload: undefined,
    existingSection: undefined,
    previousResults: undefined,
    sharedAnalysisResults: undefined,
    targetSection: undefined
  }
}

describe('Prompt builders', () => {
  it('creates clarification prompt snapshot', () => {
    const prompt = createClarificationPrompt(
      'Build a PRD for a mobile app but I only have a rough idea so far.'
    )
    expect(prompt).toMatchSnapshot()
  })

  it('creates context analysis prompt snapshot', () => {
    const prompt = createContextAnalysisPrompt('We need live analytics for enterprise clients.', {
      categorizedContext: [
        {
          id: 'ctx-1',
          category: 'constraint',
          title: 'Compliance',
          content: 'Must retain data for 24 months to satisfy enterprise audit needs.',
          priority: 'high',
          isActive: true
        },
        {
          id: 'ctx-2',
          category: 'requirement',
          title: 'Scalability',
          content: 'Handle 5k concurrent viewers without latency spikes.',
          priority: 'medium',
          isActive: true
        }
      ]
    })
    expect(prompt).toMatchSnapshot()
  })

  it('creates section detection prompt snapshot', () => {
    const prompt = createSectionDetectionPrompt(
      'Tighten our constraints for compliance and add engagement metrics.',
      existingPrdSample
    )
    expect(prompt).toMatchSnapshot()
  })

  it('creates solution prompt snapshot', () => {
    const prompt = createSolutionSectionPrompt(
      baseInput,
      contextAnalysisSample
    )
    expect(prompt).toMatchSnapshot()
  })

  it('creates key features prompt snapshot', () => {
    const promptInput: SectionWriterInput = {
      ...baseInput,
      context: {
        ...baseInput.context,
        existingSection: {
          keyFeatures: [
            'Automated anomaly detection alerts',
            'Executive digest summarizing weekly performance'
          ]
        }
      }
    }

    const prompt = createKeyFeaturesSectionPrompt(
      promptInput,
      contextAnalysisSample,
      ['Automated anomaly detection alerts']
    )
    expect(prompt).toMatchSnapshot()
  })

  it('creates success metrics prompt snapshot', () => {
    const promptInput: SectionWriterInput = {
      ...baseInput,
      context: {
        ...baseInput.context,
        existingSection: {
          successMetrics: [
            {
              metric: 'Dashboard adoption',
              target: '75% of eligible users log in weekly',
              timeline: '90 days post-launch'
            }
          ]
        }
      }
    }

    const prompt = createSuccessMetricsSectionPrompt(
      promptInput,
      contextAnalysisSample
    )
    expect(prompt).toMatchSnapshot()
  })

  it('creates constraints prompt snapshot', () => {
    const prompt = createConstraintsSectionPrompt(
      baseInput,
      contextAnalysisSample,
      ['Must comply with SOC 2 Type II'],
      ['Data warehouse maintains <5 minute latency']
    )
    expect(prompt).toMatchSnapshot()
  })

  it('creates target users prompt snapshot', () => {
    const promptInput: SectionWriterInput = {
      ...baseInput,
      context: {
        ...baseInput.context,
        existingSection: [
          'Regional operations managers overseeing daily retail performance'
        ]
      }
    }

    const prompt = createTargetUsersSectionPrompt(
      promptInput,
      contextAnalysisSample,
      ['Regional operations managers overseeing daily retail performance']
    )
    expect(prompt).toMatchSnapshot()
  })
})
