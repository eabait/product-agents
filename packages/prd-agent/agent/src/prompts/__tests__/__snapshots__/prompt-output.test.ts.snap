// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt builders creates clarification prompt snapshot 1`] = `
"You evaluate whether the following product request has enough detail to generate a meaningful PRD.

Request: "Build a PRD for a mobile app but I only have a rough idea so far."

Determine if clarification is required:
- Minimum viable information covers the product concept, primary user type, and essential functionality.
- Prefer making reasonable industry assumptions over asking obvious follow-up questions (e.g., payments for e-commerce, task tracking for project tools).

Questioning strategy:
- Ask only critical questions when a core concept, users, or primary functionality is missing.
- Optional clarifications should be limited to high-impact business context, constraints, or success criteria.

Confidence guidelines (0-100):
- 80-100 → proceed without clarification.
- 60-79 → proceed but include up to two critical questions.
- 40-59 → ask targeted questions that resolve fundamental gaps.
- 0-39 → request clarification before drafting the PRD.

Return JSON with keys:
- needsClarification (boolean)
- confidence (number 0-100)
- missingCritical (array of strings describing missing essentials)
- questions (array of critical questions only)

Output rules:
- Return strict JSON only. No prose, explanations, or code fences."
`;

exports[`Prompt builders creates constraints prompt snapshot 1`] = `
"You are a product manager refining the Constraints section of a PRD. Apply minimal edits while preserving strong existing content.

## User Input
Add real-time analytics for enterprise customers and update the launch strategy.
## Analysis Summary
- Themes: - Analytics enablement; - Enterprise readiness
- Functional requirements: - Stream KPIs in real time; - Provide historical comparisons
- Technical requirements: - Integrate with Snowflake warehouse
- UX requirements: - Accessible dashboards meeting WCAG 2.1 AA
- Constraints: - Must comply with SOC 2 Type II

## Existing Constraints
- Dependent on data warehouse refresh time under five minutes

## Existing Assumptions
- Data engineering team maintains ingestion pipeline health

## Current Constraints
- Must comply with SOC 2 Type II

## Current Assumptions
- Data warehouse maintains <5 minute latency

## Instructions
- Produce 2-6 constraints (non-negotiable limits) and 2-5 assumptions (items to validate).
- Keep statements specific, actionable, and ≤ 200 characters.
- When modifying or removing existing items, reference them directly.
- Incorporate insights from the analysis summary and provided context.

## Output JSON
- mode: "smart_merge" | "append" | "replace"
- constraints.operations: array of { action, reference, value?, rationale? } for targeted edits
- constraints.proposed: array of fully written constraints to add when appending
- assumptions.operations: same structure as constraints.operations
- assumptions.proposed: array of fully written assumptions for append actions
- summary: optional string explaining notable decisions

## Output Rules
- Return strict JSON only. No prose or commentary.
- Do not wrap the JSON in code fences or additional text."
`;

exports[`Prompt builders creates context analysis prompt snapshot 1`] = `
"You analyze the product request and extract planning signals for a PRD.

## User Input
We need live analytics for enterprise clients.

## Provided Context
- Constraint: Compliance — Must retain data for 24 months to satisfy enterprise audit needs.
- Requirement: Scalability — Handle 5k concurrent viewers without latency spikes.

## Tasks
- Themes: identify the main product goals or focus areas (string array).
- Requirements: capture functional, technical, and UX requirements as arrays of concise statements.
- Epic stories: include 3-8 epic level user stories (title + description) when possible.
- MVP features: list essential launch capabilities.
- Constraints: highlight technical, business, or compliance limitations.

## Output JSON Shape
- themes: string[]
- requirements: { functional: string[]; technical: string[]; user_experience: string[]; epics?: { title: string; description: string }[]; mvpFeatures?: string[] }
- constraints: string[]

## Output Rules
- Use empty arrays when information is unavailable.
- Keep entries concise (≤ 200 characters).
- Return strict JSON only with the keys above. No additional commentary or code fences."
`;

exports[`Prompt builders creates key features prompt snapshot 1`] = `
"You are a product manager updating the Key Features section of a PRD. Honor strong existing items and make the smallest necessary changes.

## User Input
Add real-time analytics for enterprise customers and update the launch strategy.
## Analysis Summary
- Themes: - Analytics enablement; - Enterprise readiness
- Functional requirements: - Stream KPIs in real time; - Provide historical comparisons
- Technical requirements: - Integrate with Snowflake warehouse
- UX requirements: - Accessible dashboards meeting WCAG 2.1 AA
- Epic stories: Real-time dashboards: Operations leaders view live performance trends.
- MVP features: - Live KPI dashboard; - Alert subscription management
- Constraints: - Must comply with SOC 2 Type II

## Existing Key Features
- Automated scheduled reports delivered via email
- Role-based dashboards for executives and managers

## Current Feature Set
- Automated anomaly detection alerts

## Instructions
- Produce 3-7 user-facing features. Each feature should include a short title and a value-focused description.
- Prioritize items that address the identified requirements, epics, and MVP scope.
- Keep language stakeholder-friendly (≤ 220 characters per feature).
- Prefer targeted updates or additions over full rewrites.

## Output JSON
- mode: "smart_merge" | "append" | "replace"
- operations: array of { action, referenceFeature?, feature?, rationale? } to describe updates/removals
- proposedFeatures: array of fully written features to append when adding new items
- summary: optional string explaining notable changes

## Output Rules
- Return strict JSON only. No prose or commentary.
- Do not wrap the JSON in code fences or additional text."
`;

exports[`Prompt builders creates section detection prompt snapshot 1`] = `
"You review an edit request and decide which PRD sections require updates.

## Available Sections
{
  "targetUsers": "Who the product is for",
  "solution": "What we are building and how",
  "keyFeatures": "Core functionality and differentiators",
  "successMetrics": "How success is measured",
  "constraints": "Technical, business, or regulatory limits"
}

## User Request
Tighten our constraints for compliance and add engagement metrics.

## Existing PRD Snapshot
- Target users:
- Operations managers at mid-market retail companies needing live visibility
- Solution overview: We currently provide scheduled analytics reports summarizing daily performance.
- Key features:
- Automated scheduled reports delivered via email
- Role-based dashboards for executives and managers
- Success metrics:
  • Weekly Active Analysts (60% of analysts log in weekly)
- Constraints:
- Dependent on data warehouse refresh time under five minutes

## Decision Process
- Select only the sections that must change to satisfy the request.
- If the request does not materially impact a section, leave it out.
- Typical mapping:
  • New or modified features → keyFeatures
  • Changes to audience/personas → targetUsers
  • Core solution changes or delivery approach → solution
  • Measurement or KPI updates → successMetrics
  • New limitations, dependencies, or compliance needs → constraints

## Output JSON Shape
- affectedSections: string[] containing section keys from the table above
- reasoning: object where each key is a section and the value explains why it changes
- confidence: "high" | "medium" | "low"

## Output Rules
- Be conservative: exclude sections unless there is a clear change driver.
- Return strict JSON only. No prose, explanations, or code fences."
`;

exports[`Prompt builders creates solution prompt snapshot 1`] = `
"You are a product manager drafting the Solution Overview section of a PRD. Provide clear direction while staying high level.

## User Input
Add real-time analytics for enterprise customers and update the launch strategy.
## Analysis Summary
- Themes: - Analytics enablement; - Enterprise readiness
- Functional requirements: - Stream KPIs in real time; - Provide historical comparisons
- Technical requirements: - Integrate with Snowflake warehouse
- UX requirements: - Accessible dashboards meeting WCAG 2.1 AA
- Constraints: - Must comply with SOC 2 Type II

## Existing Solution Snapshot
- Overview: We currently provide scheduled analytics reports summarizing daily performance.
- Approach: Deliver enhancements in three phases aligned to quarterly releases.

## Instructions
- Solution overview: describe WHAT is being built in 2 concise paragraphs or fewer.
- Approach: outline HOW the team will execute (architecture, phases, or methodology) in 1 paragraph.
- Address the identified requirements and constraints while keeping stakeholder-friendly language.
- Avoid low-level implementation details or open questions.

## Output JSON
- solutionOverview: string (120-600 characters, stakeholder friendly)
- approach: string (90-400 characters describing execution strategy)

## Output Rules
- Return strict JSON only. No prose or commentary.
- Do not wrap the JSON in code fences or additional text."
`;

exports[`Prompt builders creates success metrics prompt snapshot 1`] = `
"You are a product manager updating the Success Metrics section of a PRD. Respect strong existing metrics and apply only necessary changes.

## User Input
Add real-time analytics for enterprise customers and update the launch strategy.
## Analysis Summary
- Themes: - Analytics enablement; - Enterprise readiness
- Functional requirements: - Stream KPIs in real time; - Provide historical comparisons
- Technical requirements: - Integrate with Snowflake warehouse
- UX requirements: - Accessible dashboards meeting WCAG 2.1 AA
- Constraints: - Must comply with SOC 2 Type II

## Existing Success Metrics
- Weekly Active Analysts — Target: 60% of analysts log in weekly (Within two quarters of launch)

## Current Metrics to Preserve
- Dashboard adoption — Target: 75% of eligible users log in weekly (90 days post-launch)

## Instructions
- Provide 2-4 outcome metrics that signal whether the product meets its goals.
- Each metric must include a measurable target and a realistic timeline.
- Balance user value and business impact; avoid vague activity metrics.
- Reference existing metrics when updating or removing items.

## Output JSON
- mode: "smart_merge" | "append" | "replace"
- operations: array of { action, referenceMetric?, metric?, target?, timeline?, rationale? }
- proposedMetrics: array of { metric, target, timeline } for new additions
- summary: optional string describing notable adjustments

## Output Rules
- Return strict JSON only. No prose or commentary.
- Do not wrap the JSON in code fences or additional text."
`;

exports[`Prompt builders creates target users prompt snapshot 1`] = `
"You are a product manager updating the Target Users section of a PRD. Preserve strong personas and apply only necessary edits.

## User Input
Add real-time analytics for enterprise customers and update the launch strategy.
## Analysis Summary
- Themes: - Analytics enablement; - Enterprise readiness
- Functional requirements: - Stream KPIs in real time; - Provide historical comparisons
- Technical requirements: - Integrate with Snowflake warehouse
- UX requirements: - Accessible dashboards meeting WCAG 2.1 AA
- Constraints: - Must comply with SOC 2 Type II

## Existing Target Users
- Operations managers at mid-market retail companies needing live visibility

## Current Personas to Respect
- Regional operations managers overseeing daily retail performance

## Instructions
- Produce 2-4 personas. Each persona should be 1-2 sentences describing who they are and their primary need.
- Focus on specific user segments that align with the themes and requirements.
- Avoid generic labels (e.g., "end users"); highlight context such as role, company size, or pain point.
- Reference existing personas when updating or removing them.

## Output JSON
- mode: "smart_merge" | "append" | "replace"
- operations: array of { action, referenceUser?, user?, rationale? } describing targeted edits
- proposedUsers: array of fully written personas to append
- summary: optional string explaining key decisions

## Output Rules
- Return strict JSON only. No prose or commentary.
- Do not wrap the JSON in code fences or additional text."
`;
